Search: Merge `search_type=count` and `size=0`.

jpountz opened the pull request:
This commit brings the benefits of the `count` search type to search requests
that have a `size` of 0:
 - a single round-trip to shards (no fetch phase)
 - ability to use the query cache

Since `count` now provides no benefits over `query_then_fetch`, it has been
deprecated.

Close #7630

colings86 commented on src/main/java/org/elasticsearch/common/lucene/Lucene.java:
```diff
@@ -348,12 +339,8 @@ public static void writeTopDocs(StreamOutput out, TopDocs topDocs, int from) thr
 
-            out.writeVInt(topDocs.scoreDocs.length - from);
-            int index = 0;
+            out.writeVInt(topDocs.scoreDocs.length);
```
Why do we not subtract `from` now?

jpountz replied:
This method was always called with `from` = 0 so I simplified it.

colings86 replied:
ok, fair enough

colings86 commented on src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -224,8 +225,8 @@ public QuerySearchResult executeScan(ShardSearchRequest request) throws Elastics
-            context.searchType(SearchType.COUNT); // move to COUNT, and then, when scrolling, move to SCAN
-            assert context.searchType() == SearchType.COUNT;
+            context.searchType(SearchType.QUERY_THEN_FETCH); // move to QUERY_THEN_FETCH, and then, when scrolling, move to SCAN
+            assert context.searchType() == SearchType.QUERY_THEN_FETCH;
```
should this also set size = 0 for this first request so that it matches the old count functionality?

colings86 commented on src/test/java/org/elasticsearch/action/search/simple-msearch1.json:
```diff
@@ -4,7 +4,7 @@
 {}
 {"query" : {"match_all" {}}}
-{"search_type" : "count"}
+{"search_type" : "dfs_query_then_fetch"}
```
should we set size to 0 here to mirror the way the tests were run before?

jpountz replied:
This was only used by a unit test to check that msearch supports overriding the search_type per request, so I think it's fine (the test only parses the request, it does not run it).

jpountz commented:
@colings86 Thanks for the review, I pushed a new commit.

colings86 commented:
@jpountz LGTM but as I am new to this area you might want to get another review?

jpountz commented:
Makes sense. @martijnvg @bleskes @kimchy Maybe one of you would be the right person to look at this PR?

martijnvg commented:
I like the fact that without any additional overhead we can rely on `size=0` to optimize the search execution. LGTM

bleskes commented on src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java:
```diff
@@ -185,8 +178,14 @@ public boolean canCache(ShardSearchRequest request, SearchContext context) {
+        if (context.size() != 0) {
+            return false;
+        }
+        // We cannot cache with DFS because results depend not only on the content of the index but also
```
the only thing I could think of here is top hits aggs relying on score. Are there other uses where this goes wrong? (o.w. I think it might be handy to add that to the comment, it's not obvious imho)

jpountz replied:
There is also an issue with search scripts since all search scripts can read the score of the current document. I'll add a comment.

bleskes commented on src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java:
```diff
@@ -185,8 +178,14 @@ public boolean canCache(ShardSearchRequest request, SearchContext context) {
+        // We cannot cache with DFS because results depend not only on the content of the index but also
+        // on the overridden statistics. So if you ran two queries on the same index with different stats
+        // (because an other shard was updated) you would get wrong results
+        if (context.searchType() != SearchType.QUERY_THEN_FETCH && context.searchType() != SearchType.QUERY_AND_FETCH) {
```
do we want to have a CACHEABLE_TYPES enum set ?

jpountz replied:
I can do it this way if you like it better.

bleskes commented on src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java:
```diff
@@ -235,10 +234,11 @@ public QuerySearchResultProvider load(final ShardSearchRequest request, final Se
             key.shard.queryCache().onHit();
+            // restore the cached query result into the context
+            final QuerySearchResult result = context.queryResult();
+            result.readFromWithId(context.id(), value.streamInput());
```
I think the point of having BytesQuerySearchResult was that we don't have to deserialize QuerySearchResult on the source shard if we get it from cache. I believe this is lost now - was it in the way of something?

jpountz replied:
I had to deserialize the shard response all the time on the shard itself in order to figure out how many hits the shard had (to know if I can close the context). So I removed BytesQuerySearchResult which was not helping anymore.

Another idea would be to store the number of hits on the side of BytesQuerySearchResult but I'm concerned the benefit/complexit ratio would not be good?

bleskes replied:
I think the only thing we need to know is whether it's safe to free the context immediately. Maybe we can add a flag on the context for that and not have to deserialize everything?

bleskes commented on src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java:
```diff
@@ -266,7 +266,6 @@ public BytesReference call() throws Exception {
             // for now, keep the paged data structure, which might have unused bytes to fill a page, but better to keep
             // the memory properly paged instead of having varied sized bytes
             BytesReference value = out.bytes();
-            assert verifyCacheSerializationSameAsQueryResult(value, context, context.queryResult());
```
I think this is still a good assertion, no?

jpountz replied:
The assertion checks that BytesQuerySearchResult serializes the same thing as the actual QuerySearchResult, but there is no BytesQuerySearchResult anymore.

bleskes replied:
fair enough

jpountz commented:
@bleskes Pushed a new commit.

bleskes commented on src/main/java/org/elasticsearch/indices/cache/query/IndicesQueryCache.java:
```diff
@@ -218,7 +221,7 @@ public boolean canCache(ShardSearchRequest request, SearchContext context) {
      * the same cache.
      */
-    public QuerySearchResultProvider load(final ShardSearchRequest request, final SearchContext context, final QueryPhase queryPhase) throws Exception {
+    public void load(final ShardSearchRequest request, final SearchContext context, final QueryPhase queryPhase) throws Exception {
```
now that it doesn't return a value, maybe rename in to loadIntoContext , or add something to the java docs to indicate where the output is put.

bleskes commented on src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -239,6 +243,7 @@ public QuerySearchResult executeScan(ShardSearchRequest request) throws Elastics
             freeContext(context.id());
             throw ExceptionsHelper.convertToRuntime(e);
         } finally {
+            context.size(originalSize);
```
out of curiosity, did you run into problems when this was not done?

jpountz replied:
Yes: the scan-related tests would fail.

bleskes commented on src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -248,7 +253,7 @@ public ScrollQueryFetchSearchResult executeScan(InternalScrollSearchRequest requ
         try {
             processScroll(request, context);
-            if (context.searchType() == SearchType.COUNT) {
+            if (context.searchType() == SearchType.QUERY_THEN_FETCH) {
```
don't we have to set `context.size(0)` as well here? In general (not related to the change) - why do we need this flip flop between search type?

jpountz replied:
This happens on subsequent scroll requests and we need the size to be greater than 0 since this is the number of documents that we fetch on each shard at this point.

If my understanding is correct, with the scan search type we need to execute a count on the first request in order to count the number of matches (so that totalHits is correct in the response). Then for each scroll we only iterate over the subset of documents that we need.

bleskes commented:
LGTM. I left some comments with questions and suggestions. I was also wondering about the implications of this stored search templates and whether we have some bwc compatibility issue there. I'm not sure we need to solve it (no one has 1TB of stored searches, yet :)) but  we have to document it if so.

kimchy commented:
it looks good, I know this change is aimed at master (2.0), I wonder if we can still support search_type count (at least on the rest layer) and convert it to size=0 and query_then_fetch?

Also, its a shame we are loosing the non deserialization aspect of the query cache, is there still a chance to keep it?

jpountz commented:
> it looks good, I know this change is aimed at master (2.0), I wonder if we can still support search_type count (at least on the rest layer) and convert it to size=0 and query_then_fetch?

Why should we keep it, it's less confusing to have a single way to not execute the fetch phase?

> Also, its a shame we are loosing the non deserialization aspect of the query cache, is there still a chance to keep it?

Is it really worth doing? When I worked on the PR it made things complicated because other code had to make sure to use the QuerySearchResult that came back from the cache instead of the one that was attached to the context. And running a few benchmarks with this PR vs. master didn't show differences.

javanna commented:
What's the status here? I think we should get it, would love to go ahead with #9117 as a next step and make the count api become a shortcut to `size=0`.

jpountz commented:
@javanna Agreed it is a good change! :) I did not forget about it, just wanted to get https://github.com/elastic/elasticsearch/pull/9595 in first, as it would get this change more coverage. I have been traveling recently but will get back to these issues soon.

Also I totally agree that the count API should become a shortcut to the search API and be documented as such!

javanna commented:
cool thanks @jpountz no pressure :)

jpountz commented:
OK, I pushed a new version of this change that keeps `search_type=count` in but deprecated using ParseField. I had to rebase since quite some changes happened since the last version of this PR. I have had a couple of reviews already and I think I addressed all concerns. So if you would like to give it another look or have concerns, please tell me. Otherwise I will merge it soon.

javanna commented on docs/reference/search/request/search-type.asciidoc:
```diff
@@ -62,16 +62,6 @@ Same as "Query Then Fetch", except for an initial scatter phase which
-request without any docs (represented in `total_hits`), and possibly,
-including aggregations as well. In general, this is preferable to the `count`
-API as it provides more options.
-
```
nitpick...maybe you wanna leave this and mark it as deprecated?

javanna commented on rest-api-spec/api/search.json:
```diff
@@ -90,7 +90,7 @@
         "search_type": {
           "type" : "enum",
-          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "count", "scan"],
+          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "scan"],
```
I think we have to put this back if we only deprecate `count` instead of removing it? clients will still need to expose it as an option

javanna commented on rest-api-spec/api/search_template.json:
```diff
@@ -48,7 +48,7 @@
         "search_type": {
           "type" : "enum",
-          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "count", "scan"],
+          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "scan"],
```
I think we have to put this back if we only deprecate `count` instead of removing it? clients will still need to expose it as an option

javanna commented on docs/reference/search/uri-request.asciidoc:
```diff
@@ -94,7 +94,7 @@ Defaults to no terminate_after.
 |`search_type` |The type of the search operation to perform. Can be
 `dfs_query_then_fetch`, `dfs_query_and_fetch`, `query_then_fetch`,
-`query_and_fetch`, `count`, `scan`. Defaults to `query_then_fetch`. See
+`query_and_fetch`, `scan`. Defaults to `query_then_fetch`. See
```
here as well, leave count but add a deprecation note?

javanna commented on rest-api-spec/api/msearch.json:
```diff
@@ -22,7 +22,7 @@
         "search_type": {
           "type" : "enum",
-          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "count", "scan"],
+          "options" : ["query_then_fetch", "query_and_fetch", "dfs_query_then_fetch", "dfs_query_and_fetch", "scan"],
```
I think we have to put this back if we only deprecate `count` instead of removing it? clients will still need to expose it as an option

javanna commented on docs/reference/search/multi-search.asciidoc:
```diff
@@ -26,13 +26,13 @@ the `query`, `aggregations`, `from`, `size`, and so on). Here is an example:
 {"index" : "test"}
 {"query" : {"match_all" : {}}, "from" : 0, "size" : 10}
-{"index" : "test", "search_type" : "count"}
+{"index" : "test", "search_type" : "dfs_query_then_fetch"}
```
why dfs instead of default here? and shouldn't we add size: 0 as replacement for search_type count ?

jpountz replied:
My understanding was the purpose of this test was just to check a different search type

javanna replied:
alright, I didn't get the context from the diff

javanna commented on docs/reference/search/multi-search.asciidoc:
```diff
@@ -26,13 +26,13 @@ the `query`, `aggregations`, `from`, `size`, and so on). Here is an example:
 
 {"query" : {"match_all" : {}}}
-{"search_type" : "count"}
+{"search_type" : "dfs_query_then_fetch"}
```
same as above

javanna commented on src/test/java/org/elasticsearch/search/CountSearchTests.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * {@link SearchType#COUNT} is deprecated but let's make sure it still works as expected.
+ */
+public class CountSearchTests extends ElasticsearchIntegrationTest {
```
+1 thanks Adrien, I was about to complain that we don't this search_type count anymore :)

jpountz replied:
> I was about to complain

Phew, that was close!

javanna replied:
one more thing (sorry!). For the language clients, I think it would be good to also have a small REST test that uses search_type count, just to verify that all of the clients (and our REST layer) still support it.

javanna commented:
I left a bunch of docs related comments, mainly around the fact that we are now deprercating search_type count instead of removing it, hence some things need to be adjusted there.

javanna commented on docs/reference/search/uri-request.asciidoc:
```diff
@@ -94,7 +94,7 @@ Defaults to no terminate_after.
 |`search_type` |The type of the search operation to perform. Can be
 `dfs_query_then_fetch`, `dfs_query_and_fetch`, `query_then_fetch`,
-`query_and_fetch`, `count`, `scan`. Defaults to `query_then_fetch`. See
+`query_and_fetch`, `scan` or `count` (deprecated). Defaults to `query_then_fetch`. See
```
I think you could do this here: deprecated[2.0,Replaced by `size: 0`]

javanna commented on rest-api-spec/test/msearch/10_basic.yaml:
```diff
@@ -33,7 +33,7 @@
           - query:
               match_all: {}
-          - search_type: count
+          - search_type: query_then_fetch
```
use `size: 0` instead of setting the default search type?

jpountz replied:
I think the purpose of this test is to check that msearch supports setting the search type?

javanna replied:
alright, thanks
